crates/blob-store/src/lib.rs

```diff
@@ -0,0 +1,199 @@
+/// [wasi-blob-store](https://github.com/WebAssembly/wasi-blob-store) interfaces, 
+#[derive(Clone, Default)]
+pub struct BlobStore {
+    implementor: BlobStoreImplementors,
+    capability_store: HashMap<String, BasicState>,
+}
+
+impl BlobStore {
+    pub fn new(implementor: String, keyvalue_store: HashMap<String, BasicState>) -> Self {
+        Self {
+            implementor: implementor.as_str().into(),
+            capability_store: keyvalue_store,
+        }
+    }
+}
+
+#[allow(clippy::large_enum_variant)]
+#[derive(Debug, Clone, Default)]
+pub enum BlobStoreImplementors {
+    #[cfg(feature = "aws_s3")]
+    S3,
+    #[default]
+    None,
+}
+
+impl From<&str> for BlobStoreImplementors {
+    fn from(s: &str) -> Self {
+        match s {
+            #[cfg(feature = "aws_s3")]
+            S3_CAPABILITY_NAME => Self::S3,
+            p => panic!(
+                "failed to match provided name (i.e., '{p}') to any known host implementations"
+            ),
+        }
+    }
+}
+
+impl Display for BlobStoreImplementors {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            #[cfg(feature = "aws_s3")]
+            Self::S3 => write!(f, "{S3_CAPABILITY_NAME}"),
+            Self::None => panic!("No implementor specified"),
+        }
+    }
+}
+
+impl_resource!(
+    BlobStore,
+    blob_store::BlobStoreTables<BlobStore>,
+    blob_store::add_to_linker,
+    BLOB_STORE_SCHEME_NAME.to_string()
+);
+
+impl BlobStore {
+    fn fetch_state(&mut self, name: &str) -> BasicState {
+        let s: String = self.implementor.to_string();
+        let state = if let Some(r) = self.capability_store.get(name) {
+            r.clone()
+        } else if let Some(r) = self.capability_store.get(&s) {
+            r.clone()
+        } else {
+            panic!(
+                "could not find capability under name '{}' for implementor '{}'",
+                name, &s
+            );
+        };
+
+        state
+    }
+}
+
+/// This is the implementation of the wit-generated BlobStore trait for the BlobStore struct.
+#[async_trait]
+impl blob_store::BlobStore for BlobStore {
+    type Container = ContainerInner;
+    type ReadStream = ReadStreamInner;
+    type WriteStream = WriteStreamInner;
+
+    async fn container_open(&mut self, name: &str) -> Result<Self::Container, Error> {
+        let state = self.fetch_state(name);
+        tracing::log::info!("Opening implementor {}", &state.implementor);
+        let inner = Self::Container::new(state.implementor.as_str().into(), &state, name).await?;
+
+        Ok(inner)
+    }
+
+    async fn container_name(&mut self, self_: &Self::Container) -> Result<String, Error> {
+        Ok(self_.implementor.name().await?)
+    }
+    async fn container_info(
+        &mut self,
+        self_: &Self::Container,
+    ) -> Result<ContainerMetadata, Error> {
+        Ok(self_.implementor.info().await?)
```

Comment:
`self_` makes this easy to mistake for `self`. maybe rename to `container` or `ctr`?

```suggestion
        ctr: &Self::Container,
    ) -> Result<ContainerMetadata, Error> {
        Ok(ctr.implementor.info().await?)
```

same goes for the below functions too

Reply:
Thanks!